src/Orleans.Runtime/Messaging/IncomingMessageAcceptor.cs

```diff
@@ -340,19 +341,22 @@ private void ProcessAccept(SocketAsyncEventArgs e, bool completedSynchronously)
                     // Prep the socket so it will reset on close
                     sock.LingerState = receiveLingerOption;
 
-                    // Add the socket to the open socket collection
-                    if (ima.RecordOpenedSocket(sock))
+                    Task.Factory.StartNew(() =>
                     {
-                        // Get the socket for the accepted client connection and put it into the 
-                        // ReadEventArg object user token.
-                        var readEventArgs = GetSocketReceiveAsyncEventArgs(sock);
+                        // Add the socket to the open socket collection
+                        if (ima.RecordOpenedSocket(sock))
+                        {
+                            // Get the socket for the accepted client connection and put it into the 
+                            // ReadEventArg object user token.
+                            var readEventArgs = GetSocketReceiveAsyncEventArgs(sock);
 
-                        StartReceiveAsync(sock, readEventArgs, ima);
-                    }
-                    else
-                    {
-                        ima.SafeCloseSocket(sock);
-                    }
+                            StartReceiveAsync(sock, readEventArgs, ima);
+                        }
+                        else
+                        {
+                            ima.SafeCloseSocket(sock);
```

Comment:
I guess it's out-of-scope for this PR, but we're moving a blocking call onto the default TaskScheduler.
`SafeCloseSocket` calls `SocketManager.CloseSocket` which calls `Socket.Disconnect`, which ideally be async.

This might not affect Orleans code too much since most of our code uses either its own Threads or a custom TaskScheduler, but it can have a detrimental effect in co-hosting scenarios.

Unless I'm mistaken about `Disconnect` blocking.